### ISSUE TITLE
Jdbc: Annotation processor doesn't detect DataType correctly

### ISSUE THREAD

**Issue by dstepanov** [micronaut-projects/micronaut-predator#9]
I'm using PostgresSQL native UUID type, somehow default Micronaut's type for UUID is String so I need to change the type:

```
    @Id
    @Column(name = "id")
    @TypeDef(type = DataType.OBJECT)
    protected UUID id = UUID.randomUUID();
```

With repository:

```
@JdbcRepository(dialect = Dialect.POSTGRES)
public interface PersonRepository extends CrudRepository<Person, UUID> {

    List<Person> findByIdIn(List<UUID> id);

}
```

It looks like the annotation processor is incorrectly creating metadata with mapped id as DataType.String instead of DataType.Object.

```
new Object[]{"typeDefs", new AnnotationValue[]{new AnnotationValue("io.micronaut.data.annotation.TypeDef", AnnotationUtil.internMapOf(new Object[]{"names", "1", "type", DataType.STRING}))}, "parameterBinding",
new AnnotationValue[]{new AnnotationValue("io.micronaut.context.annotation.Property", AnnotationUtil.internMapOf(new Object[]{"name", "1", "value", "id"}))}, "resultDataType", DataType.ENTITY, "idType", "java.util.UUID" ...
```

**Comment by dstepanov** [comment]
I think I have found a problem in `RepositoryTypeElementVisitor#addParameterTypeDefinitions`.
Type definitions are extracted only using the method's parameter, the correct way would be to find the referenced entity field and extract `@TypeDef` from it. I was trying to come up with a solution but it looks like there is no mapping: `methodParameterName => sourcePersistentEntityPropertyName`.

I have also tried to annotate `List` with `@TypeDef` and it didn't work.
